pkg/server/compression/decompressor.go

```diff
@@ -0,0 +1,67 @@
+package compression
+
+import (
+	"compress/bzip2"
+	"compress/gzip"
+	"fmt"
+	"io"
+
+	"github.com/andybalholm/brotli"
+	"github.com/datadog/zstd"
+	"github.com/pierrec/lz4"
+	"github.com/ulikunitz/xz"
+	"github.com/ulikunitz/xz/lzma"
+)
+
+// NewDecompressor decompresses contents from an io.Reader
+// The compression type needs to be specified upfront.
+// It's the callers responsibility to close the reader when done.
+func NewDecompressor(r io.Reader, compressionType string) (io.ReadCloser, error) {
+	// Nix seems to support the following compressions:
+	// - none
+	// - br
+	// - bzip2, compress, grzip, gzip, lrzip, lz4, lzip, lzma, lzop, xz, zstd (via libarchive)
+	switch compressionType {
+	case "none":
+		return io.NopCloser(r), nil
```

Comment:
I wonder if it would make sense to forward the Close to the original r and make it a io.ReadCloser. There isn't much value of having all of them being NopCloser. Alternatively, return a io.Reader and wrap the NopCloser outside.

Reply:
The zstd reader wants to be closed, which is why this returns a `io.ReadCloser`, and not just a `io.Reader`. The others are just `io.Reader` (and don't expose a `Close` method).

I wanted to hide the complexities about which compressor wants to be closed, so expose them consistently as a `io.ReadCloser`.